components/script/script_module.rs

```diff
@@ -0,0 +1,1313 @@
+                        Some(ModuleHandler::new(Box::new(
+                            task!(all_fetched_resolve: move || {
+                                println!("promise all fetched");
+                                let promise = resolve_promise.root();
+                                promise.resolve_native(&());
+                            }),
+                        ))),
+                        Some(ModuleHandler::new(Box::new(
+                            task!(all_failure_reject: move || {
+                                println!("promise all failed");
+                                let promise = reject_promise.root();
+                                promise.reject_native(&());
+                            }),
+                        ))),
+                    );
+
+                    promise_all.append_native_handler(&handler);
+
+                    return Some(promise_all);
+                }
+            }
+        },
+        Err(err) => {
+            module_tree.set_error(Some(err));
+        },
+    }
+
+    None
+}
+
+#[allow(unsafe_code)]
+/// https://html.spec.whatwg.org/multipage/#fetch-the-descendants-of-a-module-script
+fn fetch_module_descendants(
+    owner: &ModuleOwner,
+    module_tree: &ModuleTree,
+    destination: Destination,
+    mut visited: HashSet<ServoUrl>,
+) -> Result<Vec<Rc<Promise>>, ModuleError> {
+    println!("Start to load dependencies of {}", module_tree.url.clone());
+
+    let global = owner.global();
+
+    module_tree.set_status(ModuleStatus::FetchingDescendants);
+
+    module_tree
+        .resolve_requested_modules(&global, &mut visited)
+        .map(|requested_urls| {
+            module_tree.append_descendant_urls(requested_urls.clone());
+
+            requested_urls
+                .iter()
+                .map(|requested_url| {
+                    println!("Requested url: {}", requested_url);
+
+                    {
+                        let module_map = global.get_module_map().borrow();
+                        let descendant_tree = module_map.get(&requested_url.clone());
+                        if let Some(module_tree) = descendant_tree {
+                            let status = module_tree.get_status();
+                            let promise = module_tree.get_promise().borrow();
+
+                            println!("See existing descendant: {}", requested_url.clone());
+
+                            return match promise.as_ref() {
+                                Some(p) => {
+                                    println!(
+                                        "Visited array of requested url {}: it's visited {:?} and status is {:?}",
+                                        requested_url.clone(),
+                                        visited.contains(&requested_url.clone()),
+                                        status.clone()
+                                    );
+
+                                    if visited.contains(&requested_url.clone()) {
```

Comment:
Will this ever be false? `resolve_requested_modules` always fills the visited map with all descendants